packages/solo/src/chain-cosmos-sdk.js

```diff
@@ -515,6 +522,7 @@ ${chainID} chain does not yet know of address ${clientAddr}${adviseEgress(
         const { outbox, ack } = ret;
         // console.debug('have outbox', outbox, ack);
         inbound(GCI, outbox, ack);
+        removeAckedFromMessagePool(ack);
```

Comment:
This seems pretty reasonable, although I'm curious about how it handles a particular case (I can't mentally simulate what this would do):
* solo kernel creates msg1+msg2 to send to chain, they get submitted in a txn
* while that's pending, solo kernel creates msg3+msg4
* first txn gets accepted

I'm not sure how the timing of A: txn gets accepted, and B: consequences of acceptance (e.g. the ack appearing in a new block) get reported. If they both happen at the same time, then the `blockNotifier` will cause the deliverator to see the ack for 1+2 first, so they'll be removed before it attempts to build a txn with 3+4, which is ideal.

But if something happens out of order (either the `blockNotifier` callback happens to try to send new messages before removing the acked ones, or if cosmos/tendermint is unable to publish the txn consequences in the next block), then we might wind up with a txn that includes `msg1+2+3+4`, even though 1+2 have already been accepted and processed. This isn't fatal, because comms/vattp will de-duplicate, but it'd cause larger txns than necessary.

Reply:
Usually the "acked first, then new messages" path occurs.  The nagling generally ensures that the solo has stopped twitching before the next acknowledgement happens (and the old messages are removed), before triggering more messages.

The "out of order" situation happens when there's a failure somewhere, whether a message was accepted into a node's mempool, but then dropped (usually because the node reboots), and then we reconnect to a different node, or other extenuating circumstance.